File: SeLeAs/Java/maximumTwin.java
class ListNode {
    int val;
    ListNode next;
    
    ListNode() {}
    
    ListNode(int val) {
        this.val = val;
    }
    
    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class maximumTwin {
    public int pairSum(ListNode head) {
        // Step 1: Find the middle of the list
        ListNode slow = head;
        ListNode fast = head;
        
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        
        // Step 2: Reverse the second half of the list
        ListNode prev = null;
        ListNode curr = slow;
        
        while (curr != null) {
            ListNode tmp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = tmp;
        }
        
        // Step 3: Calculate max twin sum
        ListNode head2 = prev; // Reversed half head
        int maxSum = 0;
        
        while (head2 != null) {
            maxSum = Math.max(maxSum, head.val + head2.val);
            head = head.next;
            head2 = head2.next;
        }
        
        return maxSum;
    }
}

File: SeLeAs/Python/LeetCode 75/maximumTwin.py
from typing import Optional

# Define ListNode
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        # Step 1: Find the middle of the list
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        # Step 2: Reverse second half of the list
        curr = slow
        prev = None
        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
        head2 = prev  # Head of the reversed second half

        # Step 3: Calculate max twin sum
        max_sum = 0
        while head2:
            max_sum = max(max_sum, head.val + head2.val)
            head = head.next
            head2 = head2.next

        return max_sum
